### ISSUE TITLE
Modelica.Mechanics.MultiBody.Frames.Internal.maxWithoutEvent

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#911]
**Reported by pharman on 22 Nov 2012 13:18 UTC**
From the documentation of Modelica.Mechanics.MultiBody.Frames.Internal.maxWithoutEvent:

_Function **maxWithoutEvent** returns the maximum of its two
input arguments. This functions is used instead of the Modelica
built-in function "max" or an if-statement with "noEvent(..)",
because Dymola prints a warning message when differentiating
in these cases. For the special cases as used in the MultiBody
library, these warning messages are irrelevant but will potentially
irritate the user. The C-function "maxWithoutEvent" and its
derivatives provided as additional functions "maxWithoutEvent_d"
and "maxWithoutEvent_dd" will not lead to such warning
messages._

`maxWithoutEvent` and it's derivative functions are not defined as C-functions. So the reference to them being such should be removed from the documentation and the commented-out `Header` annotation should be removed. There shouldn't be implementations of functions just to avoid warning messages in particular tools.

Is the `Integer dummy;` just a fudge to prevent the automatic inlining of the function?

Would `annotation(Inline=false, GenerateEvents=false);` not achieve the same goal (with `dummy` removed)?

```
function maxWithoutEvent
      "Maximum of the input arguments, without event and without warning message when differentiating"
      extends Modelica.Icons.Function;
      input Real u1;
      input Real u2;
      output Real y;
      //  annotation (Header="#include \"MultiBody.h\"");
    protected
      Integer dummy;
    algorithm
      y := if u1 > u2 then u1 else u2;
      dummy := 0;
      annotation (Inline=false,
    derivative=maxWithoutEvent_d);
end maxWithoutEvent;
```

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/911

**Comment by modelica-trac-importer** [comment]
**Comment by pharman on 23 Nov 2012 11:27 UTC**
(Put it in the wrong milestone, changing to MSL3.2.1)

**Comment by modelica-trac-importer** [comment]
**Comment by pharman on 9 Dec 2012 11:51 UTC**
Unless there are any objections I will replace these functions with the following:

```
    function maxWithoutEvent
      "Maximum of the input arguments without events"
      extends Modelica.Icons.Function;
      input Real u1;
      input Real u2;
      output Real y;
    algorithm
      y := if u1 > u2 then u1 else u2;
      annotation (Inline=false, GenerateEvents=false,
    derivative=maxWithoutEvent_d, Documentation(info="<html>
<p>
Function <b>maxWithoutEvent</b> returns the maximum of its two
input arguments, ensuring no events are generated by calls to
this function or derivatives of this function.
</p>
</html>"));
    end maxWithoutEvent;

    function maxWithoutEvent_d
      "First derivative of function maxWithoutEvent(..)"
      extends Modelica.Icons.Function;
      input Real u1;
      input Real u2;
      input Real u1_d;
      input Real u2_d;
      output Real y_d;
    algorithm
      y_d := if u1 > u2 then u1_d else u2_d;
      annotation (Inline=false, GenerateEvents=false, derivative(order=2) = maxWithoutEvent_dd);
    end maxWithoutEvent_d;

    function maxWithoutEvent_dd
      "First derivative of function maxWithoutEvent_d(..)"
      extends Modelica.Icons.Function;
      input Real u1;
      input Real u2;
      input Real u1_d;
      input Real u2_d;
      input Real u1_dd;
      input Real u2_dd;
      output Real y_dd;
    algorithm
      y_dd := if u1 > u2 then u1_dd else u2_dd;
      annotation(Inline=false, GenerateEvents=false);
    end maxWithoutEvent_dd;

```

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 9 Dec 2012 14:18 UTC**
Replying to [ticket:911 pharman]:

> From the documentation of Modelica.Mechanics.[wiki:MultiBody].Frames.Internal.maxWithoutEvent:
> 
> _Function **maxWithoutEvent** returns the maximum of its two
> input arguments. This functions is used instead of the Modelica
> built-in function "max" or an if-statement with "noEvent(..)",
> because Dymola prints a warning message when differentiating
> in these cases. For the special cases as used in the MultiBody
> library, these warning messages are irrelevant but will potentially
> irritate the user. The C-function "maxWithoutEvent" and its
> derivatives provided as additional functions "maxWithoutEvent_d"
> and "maxWithoutEvent_dd" will not lead to such warning
> messages._
> 
> `maxWithoutEvent` and it's derivative functions are not defined as C-functions. So the reference to them being such should be removed from the documentation and the commented-out `Header` annotation should be removed. There shouldn't be implementations of functions just to avoid warning messages in particular tools.
> 
> Is the `Integer dummy;` just a fudge to prevent the automatic inlining of the function?
> 
> Would `annotation(Inline=false, GenerateEvents=false);` not achieve the same goal (with `dummy` removed)?
> 

As a short summary:
You are right that the documentation of the function was not correct. Furthermore, it is possible to implement the function more efficiently now. Both issues have been fixed in 129c28a80ce7771f29405d808b28186874650db5.

Longer explanation:
This is a very old function and it has been updated a few times once new features in Modelica became available:

 1. The function is needed in such a way that index reduction takes place and the function will be differentiated twice. Differentiating the function correctly will result in Dirac impulses. Therefore, every reasonable tool should print a warning message (or even an error message, if the discontinuity is already detected at compile time), since tools usually cannot handle Dirac impulses. In the comment this was wrongly stated just for Dymola.
 1. However, the function is used in the MultiBody library just as a guard to avoid a division by zero. For this reason, the correct mathematical differentiation can be replaced by a version where the Dirac impulses are removed. There is no way in Modelica to state this directly to a Modelica tool. For this reason, in the beginning the function was implemented as an external C-function.
 1. Later it was recognized that this can also be formulated with pure Modelica functions and then it was forgotten to correct the documentation. 
Additionally, Dymola had the feature that it inlined automatically all functions consisting of one statement and then the technique would not work. For this reason, a local Dummy variable was introduced in the function to have two statements in the function.
 1. Since some time, the annotation "Inline=false/true" was introduced in Modelica. By setting "Inline=false" it is possible to state that the function shall not be inlined and then the Dummy equation is no longer necessary. In the fix above, the Dummy variable/equation was removed.
 1. Since some time, there is also the annotation "LateInline" available, stating that the function is inlined after all symbolic processing is performed. This option is activated in the above fix, in order that the unnecessary function call overhead is removed.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 9 Dec 2012 14:41 UTC**
Additional fix in e3f162f22fcbc912230de11dd3dde57a16610937:
Removed LateInline=true, since then events would be generated, but this would be against the feature to guard against division by zero. This could be fixed by using the new annotation GenerateEvents=false, but this annotation is not yet supported by tools.
